v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/avro/GenericRecordTypeConvertor.java

```diff
@@ -107,9 +134,99 @@ public Map<String, Value> transformChangeEvent(GenericRecord record, String srcT
+      Map<String, Value> result, GenericRecord record, String srcTableName)
+      throws InvalidTransformationException {
+    if (customTransformer == null) {
+      return result;
+    }
+    LOG.debug("Populating custom transformation for table {}: {}", srcTableName, result);
+    String spannerTableName = schemaMapper.getSpannerTableName(namespace, srcTableName);
+    // TODO: verify if direct to object (Current) works the same as Object -> JsonNode-> Object
+    // (Live).
+    Map<String, Object> sourceRowMap = genericRecordToMap(record);
+    MigrationTransformationResponse migrationTransformationResponse =
+        getCustomTransformationResponse(sourceRowMap, srcTableName, shardId);
+    if (migrationTransformationResponse.isEventFiltered()) {
+      return null;
+    }
+    Map<String, Object> transformedCols = migrationTransformationResponse.getResponseRow();
+    for (Map.Entry<String, Object> entry : transformedCols.entrySet()) {
+      LOG.debug(
+          "Updating record with {} from custom transformations for table {}", entry, srcTableName);
+      String spannerColName = entry.getKey();
+      Type spannerType =
+          schemaMapper.getSpannerColumnType(namespace, spannerTableName, spannerColName);
+      Value val =
+          getSpannerValueFromObject(
+              entry.getValue(), CUSTOM_TRANSFORMATION_AVRO_SCHEMA, spannerColName, spannerType);
+      result.put(spannerColName, val);
+    }
+    LOG.debug("Updated record with custom transformations for table {}: {}", srcTableName, result);
+    return result;
+  }
+
+  /** Converts a generic record to a Map. */
+  private Map<String, Object> genericRecordToMap(GenericRecord record) {
+    Map<String, Object> map = new HashMap<>();
+    Schema schema = record.getSchema();
+
+    for (Schema.Field field : schema.getFields()) {
+      String fieldName = field.name();
+      Object fieldValue = record.get(fieldName);
+      Schema fieldSchema = filterNullSchema(field.schema());
+      // Handle logical/record types.
+      fieldValue = handleComplexAvroTypes(fieldValue, fieldSchema, fieldName);
+      map.put(fieldName, fieldValue);
+    }
+    return map;
+  }
+
+  @VisibleForTesting
+  private MigrationTransformationResponse getCustomTransformationResponse(
+      Map<String, Object> sourceRecord, String tableName, String shardId)
+      throws InvalidTransformationException {
+
+    org.joda.time.Instant startTimestamp = org.joda.time.Instant.now();
+    MigrationTransformationRequest migrationTransformationRequest =
+        new MigrationTransformationRequest(tableName, sourceRecord, shardId, "INSERT");
+    LOG.debug(
+        "using migration transformation request {} for table {}",
+        migrationTransformationRequest,
+        tableName);
+    MigrationTransformationResponse migrationTransformationResponse =
+        customTransformer.toSpannerRow(migrationTransformationRequest);
+    org.joda.time.Instant endTimestamp = org.joda.time.Instant.now();
```

Comment:
The metrics should be in a try - finally block to ensure they are always captured even there is an exception